File: apps/bank/verify-transaction/js/verify_transaction.js
const express = require('express');
const axios = require('axios');

const app = express();
app.use(express.json());

// Required environment variables
const DIRECTUS_URL = process.env.DIRECTUS_URL;
const DIRECTUS_TOKEN = process.env.DIRECTUS_TOKEN;
const K_SINK = process.env.K_SINK;
const OTP_SECRET = process.env.OTP_SECRET;
const MAX_AMOUNT = process.env.MAX_AMOUNT ? parseFloat(process.env.MAX_AMOUNT) : 0;

const HEADERS_REMOVE = [
  'ce-id',
  'ce-specversion',
  'ce-type',
  'ce-source',
  'content-type',
  'host',
  'x-k-sink'
];

if (!DIRECTUS_URL || !DIRECTUS_TOKEN || !OTP_SECRET || MAX_AMOUNT === 0) {
  console.error('Missing required environment variables: DIRECTUS_URL, DIRECTUS_TOKEN, MAX_AMOUNT, OTP_SECRET');
  process.exit(1);
}
if (!K_SINK) {
  console.warn('Warning: K_SINK is not set. Events will not be forwarded.');
}

/**
 * Retrieves a user ID from Directus given a username.
 */
async function getUserId(username, req_id) {
  if (!username) return null;
  const url = `${DIRECTUS_URL}/items/users?filter[username][_eq]=${username}`;
  try {
    const res = await axios.get(url, {
      headers: {
        'Authorization': `Bearer ${DIRECTUS_TOKEN}`,
        'Ce-Id': req_id
      }
    });
    const data = res.data;
    if (data && data.data && data.data.length > 0) {
      return data.data[0].id;
    }
    return null;
  } catch (error) {
    return null;
  }
}

/**
 * Forwards the transaction payload to the external broker and responds to the original requester.
 * Outgoing headers are constructed to prevent overrides by incoming headers.
 */
async function forwardToBroker(req, res, payload, proceed, message, statusCode) {
  // Copy allowed incoming headers (skip ones in HEADERS_REMOVE)
  const headers = {};
  for (const key in req.headers) {
    if (!HEADERS_REMOVE.includes(key)) {
      headers[key] = req.headers[key];
    }
  }

  // Set additional headers
  headers['Ce-Id'] = req.headers['ce-id'] || '';
  headers['Ce-Specversion'] = '1.0';
  headers['Ce-Type'] = 'transaction';
  headers['Ce-Source'] = 'verify-transaction';
  headers['Ce-Dt'] = proceed ? 'true' : 'false';
  headers['Content-Type'] = 'application/json';

  // Determine target sink from header or environment variable
  const targetSink = req.headers['x-k-sink'] || K_SINK; 

  if (targetSink) {
    try {
      let bodyString = JSON.stringify(payload);
      headers['Content-Length'] = Buffer.byteLength(bodyString, 'utf8');
      await axios.post(targetSink, payload, { headers });
      console.log('Successfully forwarded event');
    } catch (error) {
      console.error('Failed to forward event:', error.message);
    }
  } else {
    console.warn('Warning: No valid sink configured. Event not forwarded.');
  }

  // If the request was not forwarded, send the response to the client
  if (!res.headersSent) {
    const responseBody = JSON.stringify({ message: message });
    res.writeHead(statusCode, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(responseBody) });
    res.end(responseBody);
  }
}

app.post('/', async (req, res) => {
  // Initialize processing state
  let proceed = true;
  let message = 'Forwarded';
  let statusCode = 200;
  
  const body = req.body;
  if (!body) {
    proceed = false;
    message = 'Invalid JSON';
    statusCode = 400;
    await forwardToBroker(req, res, {}, proceed, message, statusCode);
    return;
  }
  
  // Parse and validate "amount"
  let amount = 0;
  if (body.hasOwnProperty('amount')) {
    if (typeof body.amount === 'number' || !isNaN(body.amount)) {
      amount = parseFloat(body.amount);
    } else {
      amount = parseFloat(String(body.amount).trim());
      if (isNaN(amount)) {
        proceed = false;
        message = 'Invalid amount provided';
        statusCode = 400;
      }
    }
  } else {
    proceed = false;
    message = 'Invalid amount provided';
    statusCode = 400;
  }
  
  const clientUsername = body.client || null;
  const destinationClient = body.destination_client || null;
  // Ensure OTP values are compared as strings
  const otp = body.otp !== undefined ? String(body.otp) : null;
  const otpSecret = String(OTP_SECRET);
  
  // Check OTP if amount exceeds MAX_AMOUNT
  if (statusCode === 200 && amount > MAX_AMOUNT && (!otp || otp !== otpSecret)) {
    proceed = false;
    message = 'OTP required or incorrect';
    statusCode = 403;
  }
  
  const req_id = req.headers['ce-id'] || '';
  
  // Retrieve user IDs sequentially:
  const fromId = await getUserId(clientUsername, req_id);
  const toId = await getUserId(destinationClient, req_id);
  
  if (statusCode === 200 && (fromId === null || toId === null)) {
    proceed = false;
    message = 'Invalid client usernames';
    statusCode = 400;
  }
  
  // Build payload by copying original body and adding extra fields
  const payload = { ...body, message, from: fromId, to: toId };
  
  await forwardToBroker(req, res, payload, proceed, message, statusCode);
});

const port = 8080;
app.listen(port, () => {
  console.log(`Server started on port ${port}`);
});

File: apps/bank/verify-transaction/go/verify_transaction.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
)

// Environment variables
var (
	DIRECTUS_URL   = os.Getenv("DIRECTUS_URL")
	DIRECTUS_TOKEN = os.Getenv("DIRECTUS_TOKEN")
	K_SINK         = os.Getenv("K_SINK")
	OTP_SECRET     = os.Getenv("OTP_SECRET")
	MAX_AMOUNT     float64
)

// Headers to remove from the incoming request when forwarding.
var HEADERS_REMOVE = []string{
	"ce-id", "ce-specversion", "ce-type", "ce-source", "content-type", "host", "x-k-sink",
}

// TransactionPayload is a generic map to hold incoming JSON data.
type TransactionPayload map[string]interface{}

func main() {
	// Ensure required environment variables exist and parse MAX_AMOUNT
	if DIRECTUS_URL == "" || DIRECTUS_TOKEN == "" || OTP_SECRET == "" {
		log.Fatal("Missing required environment variables: DIRECTUS_URL, DIRECTUS_TOKEN, OTP_SECRET")
	}
	maxAmountStr := os.Getenv("MAX_AMOUNT")
	if maxAmountStr == "" {
		log.Fatal("Missing required environment variable: MAX_AMOUNT")
	}
	var err error
	MAX_AMOUNT, err = strconv.ParseFloat(maxAmountStr, 64)
	if err != nil || MAX_AMOUNT == 0 {
		log.Fatal("Invalid MAX_AMOUNT")
	}
	if K_SINK == "" {
		log.Println("Warning: K_SINK is not set. Events will not be forwarded.")
	}

	http.HandleFunc("/", handler)
	log.Println("Server started on port 8080")
	log.Fatal(http.ListenAndServe(":8080", nil))
}

func handler(w http.ResponseWriter, r *http.Request) {
	// Only allow POST requests
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		json.NewEncoder(w).Encode(map[string]string{"message": "Method Not Allowed"})
		return
	}

	// Read and parse the JSON body
	bodyBytes, err := ioutil.ReadAll(r.Body)
	if err != nil {
		respond(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	defer r.Body.Close()

	var body TransactionPayload
	if err := json.Unmarshal(bodyBytes, &body); err != nil {
		respond(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	// Initialize state variables
	proceed := true
	message := "Forwarded"
	statusCode := 200

	// Parse and validate "amount"
	var amount float64
	if amt, ok := body["amount"]; ok {
		switch v := amt.(type) {
		case float64:
			amount = v
		case string:
			// Trim and parse
			v = strings.TrimSpace(v)
			amount, err = strconv.ParseFloat(v, 64)
			if err != nil {
				proceed = false
				message = "Invalid amount provided"
				statusCode = 400
			}
		default:
			proceed = false
			message = "Invalid amount provided"
			statusCode = 400
		}
	} else {
		proceed = false
		message = "Invalid amount provided"
		statusCode = 400
	}

	// Retrieve client usernames and OTP from the payload
	clientUsername, _ := body["client"].(string)
	destinationClient, _ := body["destination_client"].(string)
	var otp string
	if val, ok := body["otp"]; ok {
		otp = fmt.Sprintf("%v", val)
	}
	// Cast OTP_SECRET to string
	otpSecret := fmt.Sprintf("%v", OTP_SECRET)

	// Check OTP if amount exceeds MAX_AMOUNT
	if statusCode == 200 && amount > MAX_AMOUNT && (otp == "" || otp != otpSecret) {
		proceed = false
		message = "OTP required or incorrect"
		statusCode = 403
	}

	// Get the request id from the headers (if any)
	reqID := r.Header.Get("Ce-Id")

	// Retrieve user IDs sequentially:
	fromID := getUserId(clientUsername, reqID)
	toID := getUserId(destinationClient, reqID)
	if statusCode == 200 && (fromID == nil || toID == nil) {
		proceed = false
		message = "Invalid client usernames"
		statusCode = 400
	}

	// Build payload: copy original payload and add extra fields
	body["message"] = message
	if fromID != nil {
		body["from"] = *fromID
	} else {
		body["from"] = nil
	}
	if toID != nil {
		body["to"] = *toID
	} else {
		body["to"] = nil
	}

	// Forward to broker
	forwardToBroker(r, body, proceed, message, statusCode)

	// Respond to the original requester
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(statusCode)
	json.NewEncoder(w).Encode(map[string]string{"message": message})
}

// getUserId makes a synchronous HTTP GET request to Directus to retrieve the user ID for a given username.
func getUserId(username, reqID string) *int {
	if username == "" {
		return nil
	}
	url := fmt.Sprintf("%s/items/users?filter[username][_eq]=%s", DIRECTUS_URL, username)
	client := &http.Client{}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil
	}
	req.Header.Set("Authorization", "Bearer "+DIRECTUS_TOKEN)
	req.Header.Set("Ce-Id", reqID)

	resp, err := client.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil
	}
	bodyBytes, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil
	}
	// Expected JSON format: { "data": [ { "id": <number> } ] }
	var result struct {
		Data []struct {
			ID int `json:"id"`
		} `json:"data"`
	}
	if err := json.Unmarshal(bodyBytes, &result); err != nil {
		return nil
	}
	if len(result.Data) > 0 {
		return &result.Data[0].ID
	}
	return nil
}

// forwardToBroker forwards the payload to the external broker (sink) with adjusted headers.
func forwardToBroker(r *http.Request, payload TransactionPayload, proceed bool, message string, statusCode int) {
	// Determine target sink from header or environment variable.
	targetSink := r.Header.Get("X-K-Sink")
	if targetSink == "" {
		targetSink = K_SINK
	}

	// Build headers: copy allowed headers from the incoming request.
	headers := make(map[string]string)
	for key, values := range r.Header {
		lowerKey := strings.ToLower(key)
		skip := false
		for _, removeKey := range HEADERS_REMOVE {
			if lowerKey == removeKey {
				skip = true
				break
			}
		}
		if !skip && len(values) > 0 {
			headers[key] = values[0]
		}
	}
	// Set additional fixed headers.
	headers["Ce-Id"] = r.Header.Get("Ce-Id")
	headers["Ce-Specversion"] = "1.0"
	headers["Ce-Type"] = "transaction"
	headers["Ce-Source"] = "verify-transaction"
	if proceed {
		headers["Ce-Dt"] = "true"
	} else {
		headers["Ce-Dt"] = "false"
	}
	headers["Content-Type"] = "application/json"

	if targetSink != "" {
		jsonPayload, err := json.Marshal(payload)
		if err != nil {
			log.Println("Error marshalling payload:", err)
			return
		}
		client := &http.Client{}
		req, err := http.NewRequest("POST", targetSink, strings.NewReader(string(jsonPayload)))
		if err != nil {
			log.Println("Error creating request to broker:", err)
			return
		}
		// Set headers on the outgoing request.
		for k, v := range headers {
			req.Header.Set(k, v)
		}
		resp, err := client.Do(req)
		if err != nil {
			log.Println("Failed to forward event:", err)
		} else {
			defer resp.Body.Close()
			log.Println("Successfully forwarded event, status code:", resp.StatusCode)
		}
	} else {
		log.Println("Warning: No valid sink configured. Event not forwarded.")
	}
}

func respond(w http.ResponseWriter, code int, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	json.NewEncoder(w).Encode(map[string]string{"message": message})
}
